### PR TITLE
Support XPS total magnetization inputs

### PR DESCRIPTION
**Pull request by cpignedoli**
## Summary

Fix XPS workflow bugs for spin-polarized systems, tagged atoms, and atom-index selection.

## Current behavior

- Spin with `tot_magnetization`: allowed without `starting_magnetization`.
- Spin with `starting_magnetization`: allowed without `tot_magnetization`; absorber magnetization is moved to `X`.
- Molecules: full core-hole final states remain charged (`tot_charge += 1`).
- Periodic systems: excited core-hole final states remain neutral (`tot_charge` unchanged).
- Atom-index mode: GUI indices are one-based; the workflow receives zero-based indices.

## Bugs fixed

- Spin/core-hole inputs now handle total magnetization and starting magnetization separately.
- Tagged absorber kinds such as `O1` are preserved for magnetization, while the chemical symbol `O` is used for pseudo/core-hole lookup.
- Atom-index submissions infer the needed core levels and write standard `symmetry_analysis_data` outputs.
- Molecule-only XPS submissions no longer build an unused relaxation step.
- The XPS result table now displays one-based site numbers.

## Code pointers

- Spin and charge handling: [`get_core_hole_inputs.py`](https://github.com/cpignedoli/aiida-qe-xspec/blob/fix/xps-spin-starting-magnetization/src/aiida_qe_xspec/workflows/functions/get_core_hole_inputs.py#L52-L150)
- Tagged molecule metadata: [`get_xspectra_structures.py`](https://github.com/cpignedoli/aiida-qe-xspec/blob/fix/xps-spin-starting-magnetization/src/aiida_qe_xspec/workflows/functions/get_xspectra_structures.py#L40-L53)
- Atom-index marked structures: [`get_marked_structures.py`](https://github.com/cpignedoli/aiida-qe-xspec/blob/fix/xps-spin-starting-magnetization/src/aiida_qe_xspec/workflows/functions/get_marked_structures.py#L31-L52)
- GUI atom-index conversion: [`workchain.py`](https://github.com/cpignedoli/aiida-qe-xspec/blob/fix/xps-spin-starting-magnetization/src/aiida_qe_xspec/gui/xps/workchain.py#L20-L49)
- XPS builder/output wiring: [`xps.py`](https://github.com/cpignedoli/aiida-qe-xspec/blob/fix/xps-spin-starting-magnetization/src/aiida_qe_xspec/workflows/xps.py#L433-L469), [`xps.py`](https://github.com/cpignedoli/aiida-qe-xspec/blob/fix/xps-spin-starting-magnetization/src/aiida_qe_xspec/workflows/xps.py#L588-L597)
- Result table site numbering: [`result.py`](https://github.com/cpignedoli/aiida-qe-xspec/blob/fix/xps-spin-starting-magnetization/src/aiida_qe_xspec/gui/xps/result/result.py#L283-L309)

## Tested on formic acid
image
image

## Co-authorship

Co-authored with OpenAI Codex.

### CODE REVIEW

File: src/aiida_qe_xspec/workflows/functions/get_core_hole_inputs.py [hunk 1]
```diff
@@ -53,14 +55,14 @@ def get_core_hole_inputs(structure, treatment, parameters, abs_site_data, **kwar
     abs_atom_marker = kwargs.get('abs_atom_marker', 'X')
     site_index = abs_site_data['site_index']
     abs_atom_kind = abs_site_data.get('kind_name', abs_site_data['symbol'])
-    updated_parameters = parameters.copy()
-    starting_mag = parameters['SYSTEM'].get('starting_magnetization', None)
-    if not starting_mag and parameters['SYSTEM'].get('nspin', None) == 2:
+    updated_parameters = deepcopy(parameters)
```
Comment (edan-bainglass) [inline src/aiida_qe_xspec/workflows/functions/get_core_hole_inputs.py]: I'm always wary of `deepcopy` ever since https://github.com/aiidalab/aiidalab-qe/pull/1341. If `parameters` carry AiiDA nodes (codes/pseudos, if we haven't popped them out by this point), `deepcopy` will trigger `Node.copy()`, which duplicates the data, but also assigns a new UUID (officially a new node). This caused issues in the past - see the above PR. Something to consider.

Comment (cpignedoli) [reply src/aiida_qe_xspec/workflows/functions/get_core_hole_inputs.py]: thanks a lot @edan-bainglass I will rework this

Comment (cpignedoli) [reply src/aiida_qe_xspec/workflows/functions/get_core_hole_inputs.py]: Thanks @edan-bainglass, this was a good point.

I replaced the recursive `deepcopy(parameters)` with targeted shallow copies only for the dictionaries that are mutated:

- top-level `parameters`
- `parameters["SYSTEM"]`
- `parameters["SYSTEM"]["starting_magnetization"]`, when present

This should avoid copying possible AiiDA nodes while still preventing mutation of the original input dictionaries.

While checking this, we also found and fixed one related edge case: XCH with AFM-like `starting_magnetization` and no `tot_magnetization` no longer tries to increment `None`. A regression test was added for that case.

I think broader cleanup of the older AFM/FM spin logic, including the hardcoded `starting_magnetization["X"] = 1` behavior, should be handled in a separate PR so this one stays focused on the current spin-input and tagged-atom bugs.

**Comment by PNOGillespie** [comment]
Hi @cpignedoli. Thanks for putting in the work to take a look at this and particularly for writing some tests (something the repo needs but I find myself without the time to get down to implementing :laughing:). I will try to find some time today and/or tomorrow to test this myself as well, particularly since I thought that I had properly implemented some of the features you're pointing to in past updates.

**Comment by cpignedoli** [comment]
Thanks a lot @PNOGillespie. One important clarification from my side: the failures I observed and the manual tests I reported were all through the AiiDAlab QE app. I did not run the workflows directly from AiiDA scripts.

The added regression tests are mostly at helper/builder level, to cover the specific input-generation issues we found while debugging the app runs.

File: src/aiida_qe_xspec/workflows/functions/get_core_hole_inputs.py (hunk 1 above)
Comment (edan-bainglass) [reply src/aiida_qe_xspec/workflows/functions/get_core_hole_inputs.py]: Thanks @cpignedoli. Note that in the referenced PR, Xing developed a routine for such "shallow-deep copies". I think it is perhaps useful to make this routine available in a more general way, perhaps in `aiida.orm.utils`. Will open an issue 🙂

Comment (edan-bainglass) [reply src/aiida_qe_xspec/workflows/functions/get_core_hole_inputs.py]: https://github.com/aiidateam/aiida-core/issues/7418 🙂

**Comment by PNOGillespie** [comment]
Hi @cpignedoli. I have run a bunch of tests myself, comparing v0.3.0 against your PR. The changes look fine to me, though I was not able to re-create the error (or, at least, what looks like an error to me) in the Formic Acid example where only 1 of the two oxygen atoms are selected for CH SCF calculations because of the O1 tag. Assuming this example was done without specifying the exact atom indices for XPS calculations, I believe this issue should already have been resolved in v0.3.0.

In any case, this PR fixes various issues and I'm happy to approve and merge it.

One thing I should say regarding your point here:
> I think broader cleanup of the older AFM/FM spin logic, including the hardcoded starting_magnetization["X"] = 1 behavior, should be handled in a separate PR so this one stays focused on the current spin-input and tagged-atom bugs.

The logic is hard-coded because (as Michael pointed out in the email chain) we can't really know the magnetic moments of the excited state *a-priori* and so have to make a guess, hoping that QE will resolve the magnetic moments correctly. I've found so far that this works well enough in all cases that I've encoutered: either the unpaired electron localizes on the absorbing atom, or it's dispersed to the rest of the structure during the SCF calculation. A more systematic approach would require a series of calculations to compute the lowest-energy configuration for the excited state: something potentially very time consuming for supercells of extended systems in particular. Indeed, it should be possible for more experienced users to do this themselves.

For insulators, and particularly AFM systems in general, where we would need to specify `starting_` and `tot_magnetization` at the same time, there is a correct way to do this detailed by Iurii Timrov in the QE source code for `hp.x` (example02, AFM NiO). The summarised steps are:

1. Run SCF calculation with small smearing term and set initial magnetization guess with `starting_magnetization` 
2. Set up new SCF calculation with fixed occupations, taking the `tot_magnetization` result from step 1 (rounded to nearest integer)
3. Run second SCF calculation to obtain the final magnetic moment. use `startingpot = file` and `startingwfc = file` to load magnetic moments from step 1 and ensure quicker SCF convergence.

I was intending to implement this process myself at some point (to be triggered only for insulators or AFM configs) and can certainly work on it if there is a strong desire for such changes.

**Comment by cpignedoli** [comment]
Hi @PNOGillespie, thank you very much for testing the PR. I am definitely in favor of keeping the behavior that has been tested and is consistent with practical simulation experience.

My comment originated from my concern that there may still be some misalignment between what the plugin GUI allows users to do and what the workflow is designed to support. If the current implementation behaves as you expect, I would not rush to take the next step. As you suggested, we should first see whether a real use case calls for a change and then act accordingly.

My understanding of the current behavior is:

**FCH/full:**

* `tot_charge` is incremented by 1.
* If `starting_magnetization` exists, the absorber kind is remapped to `X`.
* If only `tot_magnetization` exists, it is preserved, and no artificial `starting_magnetization` is created.

**XCH/excited:**

* `tot_charge` remains neutral.
* If only `tot_magnetization` exists, the current code increments it by 1.
* If a non-AFM `starting_magnetization` exists, the current code still sets `starting_magnetization["X"] = 1`.
* If AFM-like starting magnetizations exist, the value and sign of the original absorber kind are copied to `X`. If `tot_magnetization` also exists, it is shifted by +1 or −1, depending on the sign of the absorber.

**Review by PNOGillespie — APPROVED** [review]
LGTM @cpignedoli , I confirm that the added test works, but the others are broken (likely have been for a while due to neglect). I'll approve anyway since this is not your fault and open an issue as a reminder. The pre-commit hooks failed for the new test, but I can fix that afterwards.

I will look into finding the time to implement a more robust way to determine the magnetic config for the excited state. Let me know also if you find any examples in XPS or XAS which deviate from the expected behaviour that we discussed here.
